### ISSUE TITLE
Question: implementing token-based MFA support

### ISSUE THREAD

**Issue by samm-git** [Tunnelblick/Tunnelblick#417]
Hi, thank you very much for the tunnelblick OpenVPN GUI, i like it a lot. I have a question - many companies today are using OTP MFA to additionally secure auth process. Usually it is used as prefix or suffix to the password, sometime with additional separator (e.g. `<password>|OTP=<pin>`). With tunnelblick its not easy to use, because you have to type password every time, together with pin.

My idea was to create a separate [optional] input field for the token + settings to add it as suffix/prefix and optional separator. Please let me know what do you think. I do not have  osx gui experience, but can try, task does not seem very complex. Please let me know is it worth to do or its easier to implement somehow else (scripts? custom plugin?)

**Comment by johnkeates** [comment]
This would be a welcome addition indeed

**Comment by jkbullard** [comment]
I don't like adding two more text fields (one for the separator and one for the OTP) plus a checkbox to indicate whether the OTP goes before or after the password. Even just an OTP field and the before/after checkbox seems ugly, and then you'd have to type any separator yourself.

And I don't like adding a second password text field and corresponding "Save password" checkbox and a checkbox to "Always show this window".

Both of those seem too complicated and not obvious ("But they only gave me one password. Which box do I type it into?" is what 90% of users would say.

SO: Would it work to have an option to display the username/password box even if you had saved the username and password? (Something like an "Always show this login window" checkbox.)

Then you could just type the OTP into the password box at the end of the existing (saved) password, or the beginning if that's where it should go. (The existing password would be shown as bullets ("•"), as if you had just typed it in.) You could include the separator in the saved password, so you wouldn't have to type it each time.

Here's how I envision that working:

The first time you try to connect, you would fill in the username and password _without the OTP_, and check the "Save username", "Save password", and "Always show this login window" checkboxes. When you click "OK" (or "Login" or whatever it is), Tunnelblick would save the username and password (without the OTP), un-check the "Save password" checkbox, and let you add the OTP. Then when you click "OK, it would try to connect, using the password+OTP. (Because "Save password" is not checked,the password+OTP would not be saved.)

Each time you try to connect after that, it would show the login window with the username and password filled in and "Save username" and "Always show this login window" checked, but with "Save password" un-checked.

To change the "base" password, you'd just type it in place of the existing password, put a check in "Save password", and click "OK, which would be like the first time you tried to connect.

**Comment by johnkeates** [comment]
The OTP implementation in OpenVPN's own system uses a post-auth control message to ask for authenticator information which then on the client side (also OpenVPN's own GUI stuff) pops up a text box. So the OTP code isn't actually appended or prepended to the password.

**Comment by hryamzik** [comment]
FYI, I've created a PR #430.

**Comment by ssbarnea** [comment]
Some OTP implementations are  just using the normal password field so the user only gets a normal password prompt, being up to him to compose the OTP, which is usually a PIN+OTP_TOKEN, where prefix is like a static password (not necessarily numeric), and OTP_TOKEN being the code generated by the 2FA device.

When it comes to automating the login process, generating credentials using a bash script is very easy and it could look like:
```
echo -e "johndoe\npin$(/usr/local/bin/oathtool --totp -b ABCDEFG)" > /Users/johndoe/.ovpn-tmp
```
For example Viscosity VPN client for MacOS can be configured to run this command before it attempts to connect and with addition of a single line to openvpn client config, you have automatic logins. Viscosity has its own bugs here as it would loose this line from its config on any upgrade, but you can restore it.
```
auth-user-pass /Users/johndoe/.ovpn-tmp
```

Sadly Tunnelbrick fails on this aspect even worse than Viscosity because
- nobody tells you where is the connect.sh script supposed to be added
- connect is called only on new connection but not on reconnects, and OTP credentials need to be refreshed before each connection attempt regardless if this is the first or a reconnection after a drop.

If you are curious about UI ways to expose this functionality, you can check https://redmine.pfsense.org/issues/8122 which is about adding the same to pfsense. Also there is the https://community.openvpn.net/openvpn/ticket/960 ticket that may worth watching or commenting on.

**Comment by johnkeates** [comment]
Most 'good' implementations don't do this anymore. Pass+PIN is replaced by normal username+pass and then the OpenVPN protocol's challenge-response system.

**Comment by jkbullard** [comment]
>
> Sadly Tunnelbrick fails on this aspect even worse than Viscosity because
>
>    - nobody tells you where is the connect.sh script supposed to be added
>    - connect is called only on new connection but not on reconnects, and
>    OTP credentials need to be refreshed before each connection attempt
>    regardless if this is the first or a reconnection after a drop.
>
> Tunnelbrick may fail, but Tunnelblick doesn't : )


   - Tunnelblick's "connected.sh" script (not "connect.sh") is documented
   along with the other scripts in Using Scripts
   <https://tunnelblick.net/cUsingScripts.html>. The scripts go into a
   Tunnelblick VPN Configuration (.tblk), which is documented in Tunnelblick
   VPN Configurations Details <https://tunnelblick.net/cPkgs.html>. Short
   and simple version: create a folder named, for example, "Foo", put
   everything in it (OpenVPN configuration file, key and certificate files,
   and scripts), rename the folder to have suffix of ".tblk" (in the example
   that would be "Foo.tblk"), and then drag and drop it onto the Tunnelblick
   icon in the menu bar to install it as a Tunnelblick configuration named
   "Foo".

   - Tunnelblick also allows a "reconnecting.sh" script which is run on
   reconnects.

**Comment by jkbullard** [comment]
Closed because static and dynamic challenge/response are implemented in 4c1dbde99314ab0b67c70f3f8e4b13572fc0ca58 and 2d9f0e4159c66288600a468c468f9729869d025a, respectively.
